crypto/fipsmodule/sha/sha3.c

```diff
@@ -202,31 +179,133 @@ int SHA3_Update(KECCAK1600_CTX *ctx, const void *data, size_t len) {
+        // |block_size| depends on the SHA3 |bit_len| output (digest) length
+        return FIPS202_Init(ctx, SHA3_PAD_CHAR, SHA3_BLOCKSIZE(bit_len), bit_len);
+  }
+  return 0;
+}
+
+int SHA3_Update(KECCAK1600_CTX *ctx, const void *data, size_t len) {
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  if (len == 0) {
+    return 1;
+  }
+
+  return FIPS202_Update(ctx, data, len);
+}
+
+// SHA3_Final should be called once to process final digest value
+// |ctx->padded| flag does not need to be updated
+int SHA3_Final(uint8_t *md, KECCAK1600_CTX *ctx) {
+  if (ctx->md_size == 0) {
+    return 1;
+  }
+
+  if (FIPS202_Finalize(md, ctx) == 0) {
+    return 0;
+  }
+
+  Keccak1600_Squeeze(ctx->A, md, ctx->md_size, ctx->block_size, ctx->padded);
+  
+  FIPS_service_indicator_update_state();
+  return 1;
+}
+
+int SHAKE_Init(KECCAK1600_CTX *ctx, size_t block_size) {
+  if (block_size == SHAKE128_BLOCKSIZE ||
+      block_size == SHAKE256_BLOCKSIZE) { 
+        ctx->padded = 0;
+        // |block_size| depends on the SHAKE security level
+        // The output length |bit_len| is initialized to 0
+        return FIPS202_Init(ctx, SHAKE_PAD_CHAR, block_size, 0);
+  }
+  return 0;
+}
+
+int SHAKE_Absorb(KECCAK1600_CTX *ctx, const void *data, size_t len) {
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  if (len == 0) {
+    return 1;
+  }
+
+  return FIPS202_Update(ctx, data, len);
+}
+
+// SHAKE_Final is a single-shot API and can be called once to finalize absorb and squeeze phases
+// |ctx->padded| restricts consecutive calls to FIPS202_Finalize
+// Function SHAKE_Squeeze should be used for incremental XOF output
+int SHAKE_Final(uint8_t *md, KECCAK1600_CTX *ctx, size_t len) {
+  ctx->md_size = len;
+  if (ctx->md_size == 0) {
+    return 1;
+  }
+
+  if (FIPS202_Finalize(md, ctx) == 0) {
+    return 0;
+  }
+
+  Keccak1600_Squeeze(ctx->A, md, ctx->md_size, ctx->block_size, ctx->padded);
+  ctx->padded = 1;
+
+  FIPS_service_indicator_update_state();
+
+  return 1;
+}
+
+// SHAKE_Squeeze can be called multiple times
+// SHAKE_Squeeze should be called for incremental XOF output
+int SHAKE_Squeeze(uint8_t *md, KECCAK1600_CTX *ctx, size_t len) {
+  ctx->md_size = len;
+
   if (ctx->md_size == 0) {
     return 1;
   }
 
   if (ctx->padded == 0) {
-    // Pad the data with 10*1. Note that |num| can be |block_size - 1|
-    // in which case both byte operations below are performed on
-    // the same byte.
-    memset(ctx->buf + num, 0, block_size - num);
-    ctx->buf[num] = ctx->pad;
-    ctx->buf[block_size - 1] |= 0x80;
-
-    if (SHA3_Absorb(ctx->A, ctx->buf, block_size, block_size) != 0) {
+    // Skip FIPS202_Finalize if the input has been padded and the last block has been processed
+    if (FIPS202_Finalize(md, ctx) == 0) {
       return 0;
     }
   }
 
-  SHA3_Squeeze(ctx->A, md, ctx->md_size, block_size, ctx->padded);
+  Keccak1600_Squeeze(ctx->A, md, len, ctx->block_size, ctx->padded);
   ctx->padded = 1;
 
   FIPS_service_indicator_update_state();
```

Comment:
I don't think we want to add the service indicator to SHAKE_Squeeze right now, because you've updated ML-DSA/ML-KEM to use SHAKE_squeeze this will mark it as approved. We will need to lock/unlock in ML-DSA/KEM and then have it update the service indicator if it is operating in an approved manor. 

Previously ML-DSA/KEM called SHA3_Update which doesn't update the indicator it was fine. In a follow up PR when you add the service indicator for SHAKE you'll need to add more tests to all users of SHAKE_Squeeze. I'm also a little surprised the existing service indicator tests for ML-KEM didn't catch this.